specs/agents/tracing-api-otel.md

```diff
@@ -0,0 +1,235 @@
+if (otel_span.remote_contex != null) {
+    span_or_transaction = createTransactionWithParent(otel_span.remote_context);
+} else if (otel_span.parent == null) {
+    span_or_transaction = createRootTransaction();
+} else {
+    span_or_transaction = createSpanWithParent(otel_span.parent);
+}
+```
+
+### Span Kind
+
+OTel spans have an `SpanKind` property ([specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)) which is close but not strictly equivalent to our definition of spans and transactions.
+
+For both transactions and spans, an optional `otel.span_kind` property will be provided by agents when set through
+the OTel API.
+This value should be stored into Elasticsearch documents to preserve OTel semantics and help future OTel integration.
+
+Possible values are `CLIENT`, `SERVER`, `PRODUCER`, `CONSUMER` and `INTERNAL`, refer to [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind) for details on semantics.
+
+When `otel.span_kind` is not provided by the agent, APM Server should infer it using the following algorithm:
+
+```javascript
+span_kind = null;
+if (isTransaction(item)) {
+    if (item.type == "messaging") {
+        span_kind = "CONSUMER";
+    } else if (item.type == "request") {
+        span_kind = "SERVER";
+    }
+} else {
+    // span
+    if (item.type == "external" || item.type == "storage" || item.type == "db") {
+        span_kind = "CLIENT";
+    }
+}
+
+if (span_kind == null) {
+    span_kind = "INTERNAL";
+}
+
+```
+
+### Attributes mapping
+
+OTel relies on key-value pairs for span attributes.
+Keys and values are protocol-specific and are defined in [semantic convention](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions) specification.
+
+In order to minimize the mapping complexity in agents, most of the mapping between OTel attributes and agent protocol will be delegated to APM server:
+- All OTel span attributes should be captured as-is and written to agent protocol.
+- APM server will handle the mapping between OTel attributes and their native transaction/spans equivalents
+- Some native span/transaction attributes will still require mapping within agents for [compatibility with existing features](#compatibility-mapping)
+
+OpenTelemetry attributes should be stored in `otel.attributes` as a flat key-value pair mapping added to `span` and `transaction` objects:
+```json
+{
+  // [...] other span/transaction attributes
+  "otel": {
+    "span_kind": "CLIENT",
+    "attributes": {
+      "db.system": "mysql",
+      "db.statement": "SELECT * from table_1"
+    }
+  }
+}
+```
+
+APM server supports the `otel.attributes` property starting with version 7.16.0.
+
+When the APM server version does not support `otel.attributes` property, agents should use `labels` as fallback with OTel attribute
```

Comment:
In my opinion we shouldn't worry about APM Server < 7.16. I think we should remove the fallback to labels. So much of our otel support relies on recent APM Server that I think trying to support older versions is not worth the effort. Thoughts?

Reply:
My confusion lies in knowing how an agent can "know" to use `otel.attributes`, vs. `labels`. I'm thinking we should send attributes solely under `otel.attributes` and offer it as a reason for users to upgrade. `7.16` is old enough (imo) that it's not too much of an ask for them to upgrade to have this feature.

Reply:
++ to keeping it simple, and just assuming 7.16+ (with docs specifying the requirement). It may require some explanation to users to start with, but that'll diminish over time.

Reply:
I'm fine with that, too. The downside is that users may not read that part of the docs and then wonder why they don't see any context/labels/attributes. But as you said, this confusion should diminish over time, can be easily solved by pointing to docs and we don't have to complicate the code.

Reply:
+1 for keeping it simple here, I think that starting the spec (and docs if not already done yet) with a strong pre-requisite on 7.16 server should be enough.

Also, optionally adding a warning on the agent side when APM-server is < 7.16 sounds a good idea.

Please :+1:  or :-1: on this comment and I'll update the spec accordingly.

Reply:
Clarification after today's meeting, actually the new strategy will likely to keep writing to tags as fallback:
- agents do not have to issue a warning as no information is dropped, it's just not stored in the "right" location, but still there
- apm-server already implements such fallback for the non-standard otel attributes (or the ones that we don't map to our own fields).

Unless someone in @elastic/apm-server sees this as a major issue, I will proceed to revert to the previous state on this within a week.

Reply:
As there was no further comment on this, I'll close the conversation for now. As a result agents have to implement the label fallback for server < 7.16.

Reply:
@SylvainJuge I missed your earlier comment. From a server perspective, I don't have any concerns. As an agent dev, I'm wondering why we would do this as I wasn't part of the meeting you mentioned. Can you summarise what was discussed, and why the fallback is needed? Is it optional?